File: lib/components/vpc.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface VpcComponentArgs {
  vpcCidr: string;
  availabilityZones: string[];
  environmentSuffix: string;
  tags: { [key: string]: string };
}

export class VpcComponent extends pulumi.ComponentResource {
  public readonly vpc: aws.ec2.Vpc;
  public readonly publicSubnets: aws.ec2.Subnet[];
  public readonly privateSubnets: aws.ec2.Subnet[];
  public readonly internetGateway: aws.ec2.InternetGateway;
  public readonly publicRouteTable: aws.ec2.RouteTable;
  public readonly privateRouteTables: aws.ec2.RouteTable[];
  public readonly vpcId: pulumi.Output<string>;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];

  constructor(
    name: string,
    args: VpcComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:network:VpcComponent', name, {}, opts);

    const defaultResourceOptions: pulumi.ResourceOptions = { parent: this };

    // VPC
    this.vpc = new aws.ec2.Vpc(
      `vpc-${args.environmentSuffix}`,
      {
        cidrBlock: args.vpcCidr,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: {
          ...args.tags,
          Name: `vpc-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    this.vpcId = this.vpc.id;

    // Internet Gateway
    this.internetGateway = new aws.ec2.InternetGateway(
      `igw-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          ...args.tags,
          Name: `igw-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    // Calculate subnet CIDRs based on the actual VPC CIDR block
    // Parse VPC CIDR to extract base network (needed for both public and private subnets)
    const vpcCidrParts = args.vpcCidr.split('/');
    const vpcBase = vpcCidrParts[0].split('.');
    const vpcPrefix = parseInt(vpcCidrParts[1]);

    // Calculate subnet size (assuming /24 subnets)
    const subnetSize = 24;

    // Public Subnets
    // Start from offset 50 to avoid conflicts with existing subnets
    // This provides a large buffer zone for commonly used lower CIDR ranges
    const publicSubnetOffset = 50;

    this.publicSubnets = args.availabilityZones.map((az, index) => {
      // Calculate subnet CIDR based on VPC CIDR
      // For /16 VPC: 10.0.0.0/16 -> subnets: 10.0.50.0/24, 10.0.51.0/24, etc.
      // Start from offset 50 to avoid conflicts with existing subnets
      let cidrBlock: string;
      if (vpcPrefix <= 16) {
        // VPC is /16 or larger, use third octet for subnet numbering
        // Start from offset 50 to avoid conflicts with existing subnets
        const thirdOctet = publicSubnetOffset + index;
        cidrBlock = `${vpcBase[0]}.${vpcBase[1]}.${thirdOctet}.0/${subnetSize}`;
      } else {
        // VPC is smaller than /16, need more complex calculation
        // For simplicity, use index in fourth octet
        const fourthOctet = (publicSubnetOffset + index) * 64; // Space subnets by 64 addresses
        cidrBlock = `${vpcBase[0]}.${vpcBase[1]}.${vpcBase[2]}.${fourthOctet}/${subnetSize}`;
      }

      return new aws.ec2.Subnet(
        `public-subnet-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: cidrBlock,
          availabilityZone: az,
          mapPublicIpOnLaunch: true,
          tags: {
            ...args.tags,
            Name: `public-subnet-${index}-${args.environmentSuffix}`,
            Type: 'public',
          },
        },
        defaultResourceOptions
      );
    });

    this.publicSubnetIds = this.publicSubnets.map(subnet => subnet.id);

    // Public Route Table
    this.publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          ...args.tags,
          Name: `public-rt-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    // Public Route to Internet
    new aws.ec2.Route(
      `public-route-${args.environmentSuffix}`,
      {
        routeTableId: this.publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: this.internetGateway.id,
      },
      defaultResourceOptions
    );

    // Associate Public Subnets with Public Route Table
    this.publicSubnets.forEach((subnet, index) => {
      new aws.ec2.RouteTableAssociation(
        `public-rta-${index}-${args.environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: this.publicRouteTable.id,
        },
        defaultResourceOptions
      );
    });

    // Private Subnets
    // Calculate subnet CIDRs based on the actual VPC CIDR block
    // Offset private subnets to avoid overlap with public subnets
    // Private subnets start at offset 100 (public subnets use 50-99)
    const privateSubnetOffset = 100;

    this.privateSubnets = args.availabilityZones.map((az, index) => {
      // Calculate subnet CIDR based on VPC CIDR
      // Private subnets start at offset 100 to avoid overlap with public subnets (50-99)
      let cidrBlock: string;
      if (vpcPrefix <= 16) {
        // VPC is /16 or larger, use third octet for subnet numbering
        // Private subnets start at offset 100 (e.g., 10.0.100.0/24, 10.0.101.0/24)
        const thirdOctet = privateSubnetOffset + index;
        cidrBlock = `${vpcBase[0]}.${vpcBase[1]}.${thirdOctet}.0/${subnetSize}`;
      } else {
        // VPC is smaller than /16, need more complex calculation
        // For simplicity, use index in fourth octet with offset
        const fourthOctet = (privateSubnetOffset + index) * 64; // Space subnets by 64 addresses
        cidrBlock = `${vpcBase[0]}.${vpcBase[1]}.${vpcBase[2]}.${fourthOctet}/${subnetSize}`;
      }

      return new aws.ec2.Subnet(
        `private-subnet-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: cidrBlock,
          availabilityZone: az,
          tags: {
            ...args.tags,
            Name: `private-subnet-${index}-${args.environmentSuffix}`,
            Type: 'private',
          },
        },
        defaultResourceOptions
      );
    });

    this.privateSubnetIds = this.privateSubnets.map(subnet => subnet.id);

    // Private Route Tables (one per AZ for potential NAT Gateway flexibility)
    this.privateRouteTables = args.availabilityZones.map((az, index) => {
      const routeTable = new aws.ec2.RouteTable(
        `private-rt-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          tags: {
            ...args.tags,
            Name: `private-rt-${index}-${args.environmentSuffix}`,
          },
        },
        defaultResourceOptions
      );

      new aws.ec2.RouteTableAssociation(
        `private-rta-${index}-${args.environmentSuffix}`,
        {
          subnetId: this.privateSubnets[index].id,
          routeTableId: routeTable.id,
        },
        defaultResourceOptions
      );

      return routeTable;
    });

    this.registerOutputs({
      vpcId: this.vpcId,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}

File: archive/cdktf-py/Pr7954/lib/vpc.py
from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.eip import Eip
from cdktf_cdktf_provider_aws.nat_gateway import NatGateway
from cdktf_cdktf_provider_aws.route_table import RouteTable, RouteTableRoute
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation


class VpcConstruct(Construct):
    def __init__(self, scope: Construct, id: str, environment_suffix: str):
        super().__init__(scope, id)

        # VPC
        self.vpc = Vpc(self, "vpc",
            cidr_block="10.0.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"financial-vpc-{environment_suffix}",
                "Environment": f"{environment_suffix}",
                "Application": "financial-transaction-platform",
                "CostCenter": "engineering"
            }
        )

        # Internet Gateway
        self.igw = InternetGateway(self, "igw",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"financial-igw-{environment_suffix}",
                "Environment": f"{environment_suffix}",
                "Application": "financial-transaction-platform",
                "CostCenter": "engineering"
            }
        )

        # Availability Zones
        azs = ["us-east-1a", "us-east-1b", "us-east-1c"]

        # Public Subnets
        self.public_subnets = []
        for i, az in enumerate(azs):
            subnet = Subnet(self, f"public_subnet_{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=True,
                tags={
                    "Name": f"financial-public-subnet-{i+1}-{environment_suffix}",
                    "Environment": f"{environment_suffix}",
                    "Application": "financial-transaction-platform",
                    "CostCenter": "engineering",
                    "Type": "public"
                }
            )
            self.public_subnets.append(subnet)

        # Private Subnets
        self.private_subnets = []
        for i, az in enumerate(azs):
            subnet = Subnet(self, f"private_subnet_{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i+10}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=False,
                tags={
                    "Name": f"financial-private-subnet-{i+1}-{environment_suffix}",
                    "Environment": f"{environment_suffix}",
                    "Application": "financial-transaction-platform",
                    "CostCenter": "engineering",
                    "Type": "private"
                }
            )
            self.private_subnets.append(subnet)

        # Public Route Table
        self.public_route_table = RouteTable(self, "public_rt",
            vpc_id=self.vpc.id,
            route=[RouteTableRoute(
                cidr_block="0.0.0.0/0",
                gateway_id=self.igw.id
            )],
            tags={
                "Name": f"financial-public-rt-{environment_suffix}",
                "Environment": f"{environment_suffix}",
                "Application": "financial-transaction-platform",
                "CostCenter": "engineering"
            }
        )

        # Associate public subnets with public route table
        for i, subnet in enumerate(self.public_subnets):
            RouteTableAssociation(self, f"public_rt_assoc_{i}",
                subnet_id=subnet.id,
                route_table_id=self.public_route_table.id
            )

        # NAT Gateways (one per AZ for high availability)
        self.nat_gateways = []
        for i, subnet in enumerate(self.public_subnets):
            eip = Eip(self, f"nat_eip_{i}",
                domain="vpc",
                tags={
                    "Name": f"financial-nat-eip-{i+1}-{environment_suffix}",
                    "Environment": f"{environment_suffix}",
                    "Application": "financial-transaction-platform",
                    "CostCenter": "engineering"
                }
            )

            nat = NatGateway(self, f"nat_gateway_{i}",
                allocation_id=eip.id,
                subnet_id=subnet.id,
                tags={
                    "Name": f"financial-nat-{i+1}-{environment_suffix}",
                    "Environment": f"{environment_suffix}",
                    "Application": "financial-transaction-platform",
                    "CostCenter": "engineering"
                }
            )
            self.nat_gateways.append(nat)

        # Private Route Tables (one per AZ)
        self.private_route_tables = []
        for i, nat in enumerate(self.nat_gateways):
            rt = RouteTable(self, f"private_rt_{i}",
                vpc_id=self.vpc.id,
                route=[RouteTableRoute(
                    cidr_block="0.0.0.0/0",
                    nat_gateway_id=nat.id
                )],
                tags={
                    "Name": f"financial-private-rt-{i+1}-{environment_suffix}",
                    "Environment": f"{environment_suffix}",
                    "Application": "financial-transaction-platform",
                    "CostCenter": "engineering"
                }
            )
            self.private_route_tables.append(rt)

            # Associate private subnet with its route table
            RouteTableAssociation(self, f"private_rt_assoc_{i}",
                subnet_id=self.private_subnets[i].id,
                route_table_id=rt.id
            )
